src/serac/physics/materials/thermal_functional_material.hpp

```diff
@@ -0,0 +1,239 @@
+// Copyright (c) 2019-2021, Lawrence Livermore National Security, LLC and
+// other Serac Project Developers. See the top-level LICENSE file for
+// details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+
+/**
+ * @file thermal_functional_material.hpp
+ *
+ * @brief The material and load types for the thermal functional physics module
+ */
+
+#pragma once
+
+#include "serac/numerics/functional/functional.hpp"
+
+/// ThermalConductionFunctional helper structs
+namespace serac::Thermal {
+
+/// Linear isotropic thermal conduction material model
+struct LinearIsotropicConductor {
+  /// Density
+  double density_ = -1.0;
+
+  /// Specific heat capacity
+  double specific_heat_capacity_ = -1.0;
+
+  /// Constant isotropic thermal conductivity
+  double conductivity_ = -1.0;
+
+  /**
+   * @brief Function defining the thermal flux (constitutive response)
+   *
+   * @tparam T1 type of the temperature (e.g. tensor or dual type)
+   * @tparam T2 type of the temperature gradient (e.g. tensor or dual type)
+   * @param temperature_gradient Gradient of the temperature (du_dx)
+   * @return The thermal flux of the material model
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE T2 operator()(T1& /* temperature */, T2& temperature_gradient) const
+  {
+    SLIC_ASSERT_MSG(conductivity_ > 0.0,
+                    "Conductivity must be positive in the linear isotropic conductor material model.");
+
+    return -1.0 * conductivity_ * temperature_gradient;
+  }
+
+  /**
+   * @brief The density (mass per volume) of the material model
+   *
+   * @tparam T1 type of the position variable
+   * @return The density
+   */
+  template <typename T1>
+  SERAC_HOST_DEVICE double density(T1& /* x */) const
+  {
+    SLIC_ASSERT_MSG(density_ > 0.0, "Density must be positive in the linear isotropic conductor material model.");
+
+    return density_;
+  }
+
+  /**
+   * @brief The specific heat capacity (heat capacity per unit mass) of the material model
+   *
+   * @tparam T1 Type of the position variable
+   * @tparam T2 Type of the temperature variable
+   */
+  template <typename T1, typename T2>
+  SERAC_HOST_DEVICE double specificHeatCapacity(T1& /* x */, T2& /* temperature */) const
+  {
+    SLIC_ASSERT_MSG(specific_heat_capacity_ > 0.0,
+                    "Specific heat capacity must be positive in the linear isotropic conductor material model.");
+
+    return specific_heat_capacity_;
+  }
```

Comment:
the type of the spatial coordinate is always just `tensor<double, dim>`. We only need `auto` or a type-template for the arguments that will be potentially differentiated with AD.

In the future, materials with manually-written derivatives (i.e. no AD) can explicitly specify the types all of the input parameters.

Reply:
Also, I'm not sure how to best use Doxygen for this kind of pattern. These materials are all implementations of a certain concept, so documenting every single material with the same boilerplate Doxygen comments seems wasteful (but I don't know an alternative).

Reply:
I left the position arguments as `auto` as I wasn't sure if that would be needed for shape optimization. Under that context, would a plain `tensor` still be good enough?

Also I agree about the doxygen. Not really sure what the best solution is, although I would prefer boilerplate to letting missing doxygen parameters through.

Reply:
I believe the spatial coordinate can still be a plain `tensor`. In the shape optimization case, you could take a derivative w.r.t. a "design displacement" field on the surface, but that's a separate quantity.